rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,499 @@
+        self.compact_phase();
+        self.free_mark_structures();
+    }
+
+    unsafe fn alloc_mark_structures(&mut self) {
+        const BITMAP_ALGNMENT: usize = 8 * WORD_SIZE as usize;
+        let mark_offset = match self.strategy {
+            Strategy::Young => self.heap.limits.last_free / BITMAP_ALGNMENT * BITMAP_ALGNMENT,
+            Strategy::Full => self.heap.limits.base,
+        };
+        let mem_size = Bytes((self.heap.limits.free - mark_offset) as u32);
+        alloc_bitmap(self.heap.mem, mem_size, mark_offset as u32 / WORD_SIZE);
+        alloc_mark_stack(self.heap.mem);
+    }
+
+    unsafe fn free_mark_structures(&mut self) {
+        free_mark_stack();
+        free_bitmap();
+    }
+
+    unsafe fn mark_phase(&mut self) {
+        self.marked_space = 0;
+        self.mark_root_set();
+        self.mark_all_reachable();
+    }
+
+    unsafe fn mark_root_set(&mut self) {
+        self.mark_static_roots();
+
+        if (*self.heap.roots.continuation_table_ptr_loc).is_ptr() {
+            self.mark_object(*self.heap.roots.continuation_table_ptr_loc);
+        }
+
+        if self.strategy == Strategy::Young {
+            self.mark_additional_young_root_set();
+        }
+    }
+
+    unsafe fn mark_static_roots(&mut self) {
+        let root_array = self.heap.roots.static_roots.as_array();
+        for i in 0..root_array.len() {
+            let object = root_array.get(i).as_obj();
+            debug_assert_eq!(object.tag(), TAG_MUTBOX);
+            debug_assert!((object as usize) < self.heap.limits.base);
+            self.mark_root_mutbox_fields(object as *mut MutBox);
+        }
+    }
+
+    unsafe fn mark_additional_young_root_set(&mut self) {
+        let mut iterator = REMEMBERED_SET.as_ref().unwrap().iterate();
+        while iterator.has_next() {
+            let location = iterator.current().get_raw() as *mut Value;
+            let value = *location;
+            // Check whether the location still refers to young object as this may have changed
+            // due to subsequent writes to that location after the write barrier recording.
+            if value.points_to_or_beyond(self.heap.limits.last_free) {
+                self.mark_object(value);
+            }
+            iterator.next();
+        }
+    }
+
+    unsafe fn mark_object(&mut self, object: Value) {
+        let pointer = object.get_ptr() as u32;
+        if self.strategy == Strategy::Young && pointer < self.heap.limits.last_free as u32 {
+            return;
+        }
+        debug_assert_eq!(pointer % WORD_SIZE, 0);
+
+        let obj_idx = pointer / WORD_SIZE;
+        if get_bit(obj_idx) {
+            return;
+        }
+        set_bit(obj_idx);
+
+        push_mark_stack(self.heap.mem, pointer as usize, object.tag());
+        self.marked_space += object_size(pointer as usize).to_bytes().as_usize();
+    }
+
+    unsafe fn mark_all_reachable(&mut self) {
+        while let Some((obj, tag)) = pop_mark_stack() {
```

Comment:
I expect its still safe to recover the tag from the object itself here. I guess a simple assert could verify that claim.

Reply:
(modulo the special treatment of arrray)

Reply:
The change was really simple. [7dd1787](https://github.com/dfinity/motoko/pull/3495/commits/7dd178754f73f21d364c56173feaa15ab557b54b)